File: lib/directorio.ts
import { apiClient } from './api';

export interface TipoPersona {
  tipo: 'natural' | 'juridica';
}

export interface TipoDocumento {
  tipo: 'DNI' | 'RUC' | 'CE' | 'PAS';
}

export interface DirectorioEntry {
  id: number;
  tipo: 'cliente' | 'juzgado' | 'especialista';
  nombre: string;
  email?: string;
  telefono?: string;
  direccion?: string;
  // Cliente fields
  tipo_persona?: 'natural' | 'juridica';
  nombres?: string;
  apellidos?: string;
  razon_social?: string;
  doc_tipo?: 'DNI' | 'RUC' | 'CE' | 'PAS';
  doc_numero?: string;
  // Juzgado fields
  distrito_judicial?: string;
  // Especialista fields
  especialidad?: string;
  numero_colegiado?: string;
  juzgado_id?: number;
  activo: boolean;
  created_at: string;
  updated_at: string;
}

export interface DirectorioCreate {
  tipo: 'cliente' | 'juzgado' | 'especialista';
  nombre: string;
  email?: string;
  telefono?: string;
  direccion?: string;
  tipo_persona?: 'natural' | 'juridica';
  nombres?: string;
  apellidos?: string;
  razon_social?: string;
  doc_tipo?: 'DNI' | 'RUC' | 'CE' | 'PAS';
  doc_numero?: string;
  distrito_judicial?: string;
  especialidad?: string;
  numero_colegiado?: string;
  juzgado_id?: number;
  activo?: boolean;
}

export interface DirectorioUpdate {
  nombre?: string;
  email?: string;
  telefono?: string;
  direccion?: string;
  tipo_persona?: 'natural' | 'juridica';
  nombres?: string;
  apellidos?: string;
  razon_social?: string;
  doc_tipo?: 'DNI' | 'RUC' | 'CE' | 'PAS';
  doc_numero?: string;
  distrito_judicial?: string;
  especialidad?: string;
  numero_colegiado?: string;
  juzgado_id?: number;
  activo?: boolean;
}

export interface EstadisticasDirectorio {
  por_tipo: Record<string, number>;
  total: number;
}

const directorioAPI = {
  // Listar con filtros
  getAll: async (skip = 0, limit = 100, tipo?: string, activosSolo = false) => {
    const params = new URLSearchParams();
    params.append('skip', skip.toString());
    params.append('limit', limit.toString());
    if (tipo) params.append('tipo', tipo);
    if (activosSolo) params.append('activos_solo', 'true');
    
    return apiClient.request<DirectorioEntry[]>({
      method: 'GET',
      url: `/directorio?${params.toString()}`,
    });
  },

  // Obtener por ID
  getById: async (id: number) => {
    return apiClient.request<DirectorioEntry>({
      method: 'GET',
      url: `/directorio/${id}`,
    });
  },

  // Listar solo clientes
  getClientes: async () => {
    return apiClient.request<DirectorioEntry[]>({
      method: 'GET',
      url: '/directorio/clientes',
    });
  },

  // Listar solo juzgados
  getJuzgados: async () => {
    return apiClient.request<DirectorioEntry[]>({
      method: 'GET',
      url: '/directorio/juzgados',
    });
  },

  // Listar solo especialistas
  getEspecialistas: async () => {
    return apiClient.request<DirectorioEntry[]>({
      method: 'GET',
      url: '/directorio/especialistas',
    });
  },

  // Listar especialistas por juzgado
  getEspecialistasByJuzgado: async (juzgadoId: number) => {
    return apiClient.request<DirectorioEntry[]>({
      method: 'GET',
      url: `/directorio/juzgados/${juzgadoId}/especialistas`,
    });
  },

  // Buscar
  search: async (query: string, tipo?: string) => {
    const params = new URLSearchParams();
    params.append('q', query);
    if (tipo) params.append('tipo', tipo);
    
    return apiClient.request<DirectorioEntry[]>({
      method: 'GET',
      url: `/directorio/buscar?${params.toString()}`,
    });
  },

  // Crear
  create: async (data: DirectorioCreate) => {
    return apiClient.request<DirectorioEntry>({
      method: 'POST',
      url: '/directorio',
      data,
    });
  },

  // Actualizar
  update: async (id: number, data: DirectorioUpdate) => {
    return apiClient.request<DirectorioEntry>({
      method: 'PUT',
      url: `/directorio/${id}`,
      data,
    });
  },

  // Eliminar
  delete: async (id: number) => {
    return apiClient.request<void>({
      method: 'DELETE',
      url: `/directorio/${id}`,
    });
  },

  // Estadísticas
  getEstadisticas: async () => {
    return apiClient.request<EstadisticasDirectorio>({
      method: 'GET',
      url: '/directorio/estadisticas',
    });
  },
};

export { directorioAPI };
export default directorioAPI;

File: backend/app/schemas/directorio.py
"""Pydantic schemas para Directorio (clientes, juzgados, especialistas unificados)"""

from pydantic import BaseModel, Field, EmailStr
from typing import Optional, List
from datetime import datetime
from enum import Enum


class TipoDirectorio(str, Enum):
    cliente = "cliente"
    juzgado = "juzgado"
    especialista = "especialista"


class TipoPersona(str, Enum):
    natural = "natural"
    juridica = "juridica"


class TipoDocumento(str, Enum):
    DNI = "DNI"
    RUC = "RUC"
    CE = "CE"
    PAS = "PAS"


# Base schema con campos comunes
class DirectorioBase(BaseModel):
    tipo: TipoDirectorio
    nombre: str = Field(..., max_length=180)
    email: Optional[EmailStr] = None
    telefono: Optional[str] = Field(None, max_length=30)
    direccion: Optional[str] = Field(None, max_length=250)
    activo: bool = True


# Campos específicos para CLIENTE
class DirectorioClienteCreate(DirectorioBase):
    tipo: TipoDirectorio = TipoDirectorio.cliente
    tipo_persona: TipoPersona
    nombres: Optional[str] = Field(None, max_length=120)
    apellidos: Optional[str] = Field(None, max_length=120)
    razon_social: Optional[str] = Field(None, max_length=180)
    doc_tipo: TipoDocumento = TipoDocumento.DNI
    doc_numero: str = Field(..., max_length=20)


# Campos específicos para JUZGADO
class DirectorioJuzgadoCreate(DirectorioBase):
    tipo: TipoDirectorio = TipoDirectorio.juzgado
    distrito_judicial: Optional[str] = Field(None, max_length=120)


# Campos específicos para ESPECIALISTA
class DirectorioEspecialistaCreate(DirectorioBase):
    tipo: TipoDirectorio = TipoDirectorio.especialista
    nombres: Optional[str] = Field(None, max_length=120)
    apellidos: Optional[str] = Field(None, max_length=120)
    especialidad: Optional[str] = Field(None, max_length=120)
    numero_colegiado: Optional[str] = Field(None, max_length=50)
    juzgado_id: Optional[int] = None  # Especialista vinculado a un juzgado


# Union type para crear cualquier tipo
class DirectorioCreate(BaseModel):
    tipo: TipoDirectorio
    nombre: str
    email: Optional[EmailStr] = None
    telefono: Optional[str] = None
    direccion: Optional[str] = None
    tipo_persona: Optional[TipoPersona] = None
    nombres: Optional[str] = None
    apellidos: Optional[str] = None
    razon_social: Optional[str] = None
    doc_tipo: Optional[TipoDocumento] = None
    doc_numero: Optional[str] = None
    distrito_judicial: Optional[str] = None
    especialidad: Optional[str] = None
    numero_colegiado: Optional[str] = None
    juzgado_id: Optional[int] = None
    activo: bool = True


class DirectorioUpdate(BaseModel):
    nombre: Optional[str] = None
    email: Optional[EmailStr] = None
    telefono: Optional[str] = None
    direccion: Optional[str] = None
    tipo_persona: Optional[TipoPersona] = None
    nombres: Optional[str] = None
    apellidos: Optional[str] = None
    razon_social: Optional[str] = None
    doc_tipo: Optional[TipoDocumento] = None
    doc_numero: Optional[str] = None
    distrito_judicial: Optional[str] = None
    especialidad: Optional[str] = None
    numero_colegiado: Optional[str] = None
    juzgado_id: Optional[int] = None
    activo: Optional[bool] = None


class DirectorioResponse(BaseModel):
    id: int
    tipo: TipoDirectorio
    nombre: str
    email: Optional[str] = None
    telefono: Optional[str] = None
    direccion: Optional[str] = None
    tipo_persona: Optional[TipoPersona] = None
    nombres: Optional[str] = None
    apellidos: Optional[str] = None
    razon_social: Optional[str] = None
    doc_tipo: Optional[TipoDocumento] = None
    doc_numero: Optional[str] = None
    distrito_judicial: Optional[str] = None
    especialidad: Optional[str] = None
    numero_colegiado: Optional[str] = None
    juzgado_id: Optional[int] = None
    activo: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# Respuesta especializada por tipo
class DirectorioClienteResponse(DirectorioResponse):
    tipo: TipoDirectorio = TipoDirectorio.cliente
    tipo_persona: TipoPersona
    doc_numero: str


class DirectorioJuzgadoResponse(DirectorioResponse):
    tipo: TipoDirectorio = TipoDirectorio.juzgado
    distrito_judicial: Optional[str] = None


class DirectorioEspecialistaResponse(DirectorioResponse):
    tipo: TipoDirectorio = TipoDirectorio.especialista
    nombres: str
    apellidos: str
